### PR TITLE
Data split for time series

### LINKED ISSUE #1381: Data splitting for time series
Requirements:

1. We're going to start with [Rolling Origin Cross Validation](https://robjhyndman.com/hyndsight/tscv/)
2. Estimators should not be scored on the dates that they are fit on.
3. The max_lag and gap values will mean some rows will be "thrown out" because of nans. Let's try to minimize this amount.

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fixes #1381.

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: evalml/automl/data_splitters/time_series_split.py
Comment (jeremyliweishih) [inline evalml/automl/data_splitters/time_series_split.py:72]: why yield instead of returning at the end?

**Review by jeremyliweishih — APPROVED** [review]
this looks great to me! I don't really have any other comments other than it would be helpful to include links to `Delayed Feature Transformer` and `TimeSeriesSplit` and also an explanation of what ROCV is. It took my a while to understand the logic of everything and it only clicked once I read through all the context.

File: evalml/tests/automl_tests/test_time_series_split.py
Comment (dsherry) [inline evalml/tests/automl_tests/test_time_series_split.py:31]: Ah neat, I didn't expect the answers here to be this cleanly defined.

Considering the case when `max_delay=0` and `gap=0` for simplicity, why does the first training split range from 10-01 to 10-10 excluding the last day (i.e. 9 days), but the test split always has 6 days, and each subsequent update increases by 7 days? I guess my question is, why aren't all these splits the same size?

I think the simplest algo to follow for this would be if we divide the data into `n_splits` equal sections, and then each split in the generator uses the i-th ordered section as the test data, and the concatenation of all previous ordered sections as the training data. It appears the algo we've currently got matches this in the ordering, but the section sizes are a bit different. How come?

**Review by dsherry — APPROVED** [review]
@freddyaboulton looks good!

I left some questions about the impl, as well as a suggested simplification (don't use sklearn splitter). I also left a question about why the algo we appear to be using isn't dividing the data into equally-sized sections. That said, the impl makes sense to me, these questions aren't blocking merge, and we can certainly file something to revisit our timeseries splitting, if you prefer to merge and address this later; your call.

File: evalml/tests/automl_tests/test_time_series_split.py
Comment (freddyaboulton) [reply evalml/tests/automl_tests/test_time_series_split.py:31]: I think the reason here is that the number of datapoints (31) is not divisible by the number of folds (4 because splitting three times creates four chunks of data). 

Splitting the data equally into four creates would results in four 7-day chunks of data with three days leftover. We just add the first three days to the first chunk of data. 

I think if the data isn't evenly divisible by the number of splits, you'll always end up in a situation where one of the splits is slightly larger (unless you throw out data which I don't think we should do).

If make a dataframe with 32 rows instead of 31 and use n_splits=3, the first train/test are 8 days long and each subsequent train set increases by 8 days.

File: evalml/automl/data_splitters/time_series_split.py
Comment (freddyaboulton) [reply evalml/automl/data_splitters/time_series_split.py:6]: Yes!

File: evalml/automl/data_splitters/time_series_split.py
Comment (freddyaboulton) [reply evalml/automl/data_splitters/time_series_split.py:71]: Yes they will but we make sure they never extend past the indices of `X`!

File: evalml/automl/data_splitters/time_series_split.py
Comment (freddyaboulton) [reply evalml/automl/data_splitters/time_series_split.py:65]: I'm not so sure. Sklearn already has a tested implementation of the rolling origin cv algorithm we set forth in the design phase so the only thing we need to take care of is that the splits are aware of  `gap` and `max_delay`.  I prefer to delegate the part we can delegate to one of our dependencies and the code that lives in evalml only needs to worry about evalml requirements. Besides, I feel like implementing rolling-origin cv ourselves would amount to something very similar to sklearn's implementation which I feel isn't very DRY . 

So I vote we keep the implementation as is for now and then we can revisit later if we decide to move away from rolling origin cv!

File: evalml/automl/data_splitters/time_series_split.py
Comment (freddyaboulton) [reply evalml/automl/data_splitters/time_series_split.py:70]: Yep `train[0] == 0` so I'll make that change. I'll also make sure the `train` doesn't extend past `X` !

File: evalml/tests/automl_tests/test_time_series_split.py
Comment (dsherry) [reply evalml/tests/automl_tests/test_time_series_split.py:40]: @freddyaboulton got it, thanks for the explanation. I wanted to make sure there wasn't a bug lurking in there somewhere. I would've guessed the underlying sklearn splits would be as even as possible but I guess not! It'll be interesting to learn how this behaves on other datasets.

**Pull request merged**
